File: lab1/kod_aplikacji/kod_aplikacji/backend/src/main/java/engineering/app/models/Book.java
package engineering.app.models;

import com.fasterxml.jackson.annotation.*;
import jakarta.persistence.*;
import lombok.*;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@Setter
@Getter

public class Book {
    @Id
    @SequenceGenerator(
            name="book_sequence",
            sequenceName = "book_sequence",
            allocationSize = 1
    )
    @GeneratedValue(
            strategy = GenerationType.SEQUENCE,
            generator = "book_sequence"
    )
    private Long id;
    private String title;

    @ManyToMany(fetch = FetchType.LAZY,
            cascade={CascadeType.MERGE}
    )
    @JoinTable(
            name = "book_author",
            joinColumns = @JoinColumn(name = "book_id"),
            inverseJoinColumns = @JoinColumn(name = "author_id"))
    private Set<Author> authors = new HashSet<>();
    private Integer pageCount;
    @ManyToOne(fetch = FetchType.LAZY, cascade = {CascadeType.MERGE})
    private Publisher publisher;
    @ManyToOne(fetch = FetchType.LAZY, cascade = {CascadeType.MERGE})
    private Category category;
    @Column(unique = true)
    private String isbn;
    @Setter
    private boolean available = true;
    @Setter
    @JsonProperty
    private boolean isPublished;
    @OneToMany(mappedBy = "book", cascade = CascadeType.ALL)
    @JsonIgnore
    private List<BookLoan> bookLoans;
    @Lob
    @Column(length=1000000)
    private String description;
    private Integer releaseYear;
    @JsonProperty
    private boolean isEbook;
    @Setter
    private boolean isDeleted;
    @OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @JoinColumn(name = "image_data_id")
    private ImageData imageData;
    @OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @JoinColumn(name = "file_data_id")
    @JsonIgnore
    private FileData fileData;
    @ManyToOne(fetch = FetchType.LAZY, cascade = {CascadeType.MERGE})
    private User user;
}

File: lab1/kod_aplikacji/kod_aplikacji/frontend/src/Models/Book.ts
import {User} from "./User";

export type Book = {
  id?: number;
  authors: Author[];
  title: string;
  pageCount: number | null;
  description: string | null;
  coverFile?: File | null | string;
  publisher: Publisher;
  category: Category;
  isbn: string;
  available: boolean;
  isPublished: boolean;
  releaseYear: number | null;
  imageData?: {
    id: number;
    imageData: string;
    name: string;
    type: string;
  };
  isEbook: boolean;
  user: User | null;
  isDeleted: boolean;
};

export type Author = {
  id?: number | null;
  firstName: string;
  lastName: string;
};
export type Category = {
  id?: number | null;
  name: string;
};
export type Publisher = {
  id?: number | null;
  name: string;
};
